Global Refactoring

brash6 opened the pull request:
I am opening this PR so that you can start to see the work I have done until now. 
This PR is still messy and not clean.

What I have done (a lot is inspired from the work initiated by @houssamzenati and what we have defined in the issue #90 : 

- Modularize estimators 
- Add nuisances functions in a nuisances folder / Add useful utils functions 
- Create an experiment.py file to test modularized estimators with the data generation from med_bench

TO DO : 

- Ensure every estimator is working as expected
- Explicitly decide which set of parameters to be used in every scenario (data generation, classifiers, regressors, every other hyper parameters)
- Create tests files covering every defined scenario
- Adapt CI to new tests files
- Add clean docstrings everywhere
- Clean files (remove unused files / functions / variables, enhance / unify naming, strongtyping, etc.)
- Add an example file in which each estimator is run


FYI @judithabk6 @bthirion @houssamzenati

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,295 @@
+        ----------
+
+        hat_tau array-like, shape (n_samples)
+                estimated risk
```
what is hat_tau here ? an estimated total effect size ?

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,295 @@
+        """
+        xm = np.hstack((x, m))
+        self._classifier_t_xm = self.classifier.fit(xm, t)
+
```
return self ? (Note sure, this is a suggestion tbd )

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,295 @@
+        Parameters
+        ----------
+        mediator_type : str
+            mediator type (binary or continuous)
```
```suggestion
            mediator type ("binary" or "continuous")
```

bthirion replied:
What about multi-dimensional ? (could be part of "continuous")

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,295 @@
+        """
+        self.rng = np.random.RandomState(123)
+
+        self.mediator_type = mediator_type
```
curation needed on mediator_type

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,295 @@
+        return score
+    
+
+    def fit_treatment_propensity_x_nuisance(self, t, x):
```
Are these functions meant to be part of the public API or private ?

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,295 @@
+
+
+    def fit_cross_conditional_mean_outcome_nuisance_discrete(self, t, m, x, y):
+        n = len(y)
```
docstring needed

bthirion commented on src/med_bench/estimation/coefficient_product.py:
```diff
@@ -0,0 +1,93 @@
+
+class CoefficientProduct(Estimator):
+
+    def __init__(self, clip : float, trim : float, regularize : bool, **kwargs):
```
Should clip and trim be provided fora ll estimators. Actually, they're not used in the CoefficientProduct

bthirion commented on src/med_bench/estimation/dml.py:
```diff
@@ -0,0 +1,181 @@
+    Parameters
+    ----------
+        settings (dict): dictionnary of parameters
+        lbda (float): regularization parameter
```
```suggestion
        alpha (float): regularization parameter
``` Note: sklearn calls regularization  parameters `alpha`

bthirion commented on src/med_bench/estimation/dml.py:
```diff
@@ -0,0 +1,181 @@
+        self._normalized = normalized
+        self._sample_split = sample_split
+
+    def resize(self, t, m, x, y):
```
Is this unique to DoubleMachineLearning ?

bthirion commented on src/med_bench/estimation/dml.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        t, m, x, y = self.resize(t, m, x, y)
+
+        p_x, p_xm = estimate_treatment_probabilities(t,
```
Just wondering why `estimate_treatment_probabilities` is not a method of `Estimator` ?

bthirion commented on src/med_bench/estimation/dml.py:
```diff
@@ -0,0 +1,181 @@
+                                                          self._classifier_t_xm)
+
+
+        mu_0mx, mu_1mx, E_mu_t0_t0, E_mu_t0_t1, E_mu_t1_t0, E_mu_t1_t1 = estimate_cross_conditional_mean_outcome_nesting(m, x, y, self.regressors)
```
Same question here.

bthirion commented on src/med_bench/estimation/dml.py:
```diff
@@ -0,0 +1,181 @@
+        eta_t1t0 = np.mean(y1m0)
+        eta_t0t1 = np.mean(y0m1)
+
+                # effects computing
```
```suggestion
        # effects computing
```

bthirion commented on src/med_bench/estimation/g_computation.py:
```diff
@@ -0,0 +1,190 @@
+        
+        if self._procedure == 'discrete':
+            if not is_array_integer(m):
+                self._bucketizer = KMeans(n_clusters=10, random_state=self.rng,
```
There are some hard-coded decision here. To be discussed.

bthirion commented on src/med_bench/estimation/g_computation.py:
```diff
@@ -0,0 +1,190 @@
+        }
+        return causal_effects
+    
+    def estimate_nesting(self, t, m, x, y):
```
to be clarified what this does

bthirion commented on src/med_bench/estimation/ipw.py:
```diff
@@ -0,0 +1,121 @@
+from med_bench.utils.decorators import fitted
+
+
+class ImportanceWeighting(Estimator):
```
```suggestion
class InversePropensityWeighting(Estimator):
```

bthirion commented on src/med_bench/estimation/mr.py:
```diff
@@ -0,0 +1,279 @@
+
+        elif self._procedure == 'discrete':
+
+            if not is_array_integer(m):
```
should rely o the `mediator_type` attribute ?

bthirion commented on src/med_bench/estimation/mr.py:
```diff
@@ -0,0 +1,279 @@
+
+        if self.verbose:
+            print("Nuisance models fitted")
+
```
A scikit learn convention is that `.fit()` methods return `self`

bthirion commented on src/med_bench/estimation/mr.py:
```diff
@@ -0,0 +1,279 @@
+
+
+        else:
+            if not is_array_integer(m):
```
see above comment on mediator_type

bthirion commented on src/med_bench/estimation/tmle.py:
```diff
@@ -0,0 +1,267 @@
+
+class TMLE(Estimator):
+    """Implementation of targeted maximum likelihood estimator
+
```
ref needed

bthirion commented on src/med_bench/estimation/tmle.py:
```diff
@@ -0,0 +1,267 @@
+        return t, m, x, y
+    
+
+    def one_step_correction_direct(self, t, m, x, y):
```
should be a private method  I think

bthirion commented on src/med_bench/experiment.py:
```diff
@@ -0,0 +1,37 @@
+from med_bench.get_simulated_data import simulate_data
+from med_bench.nuisances.utils import _get_regularization_parameters
+from med_bench.utils.constants import CV_FOLDS
+
```
Is this meant to be an example ?

bthirion commented on src/med_bench/nuisances/density.py:
```diff
@@ -0,0 +1,315 @@
+        probabilities f(M|T=0,X)
+    f_m1x, array-like, shape (n_samples)
+        probabilities f(M|T=1,X)
+    """
```
The functions in these modules probably need full docstrings

bthirion commented on src/med_bench/nuisances/density.py:
```diff
@@ -0,0 +1,315 @@
+            ]
+        return kernel_density_list
+
+    def pdf(self, y, x):
```
what dpes pdf mean here ?

bthirion reviewed (commented):
Looks like a big step.
I just provided a few thoughts. Hope it helps.

judithabk6 commented on src/med_bench/estimation/mr.py:
```diff
@@ -0,0 +1,279 @@
+import numpy as np
```
not sure about all the file names with abbreviations, should be named `multiply_robust.py`?, also do we want to group some estimators together? as `MultiplyRobust` and `DoubleMachineLearning`are very close? should be discussed further

judithabk6 commented on src/med_bench/estimation/tmle.py:
```diff
@@ -0,0 +1,267 @@
+ALPHA=10
+
+from med_bench.utils.decorators import fitted
+from med_bench.utils.utils import _get_interactions
```
not sure we need interactions here, because we can use non-linear methods (and maybe some equivalent of the SuperLearner at some point), the place where interactions are key is for the coefficient product estimator

judithabk6 commented on src/med_bench/experiment.py:
```diff
@@ -0,0 +1,37 @@
+    coef_prod_estimator.fit(t_train, m_train, x_train, y_train)
+    causal_effects = coef_prod_estimator.estimate(t_test, m_test, x_test, y_test)
+
+    r_risk_score = coef_prod_estimator.score(t_test, m_test, x_test, y_test, causal_effects['total_effect'])
```
should be implemented in utils maybe, with a metrics module. Just saw that it existed, so need to import from there. Not sure about the naming scores vs metrics

judithabk6 commented on src/med_bench/nuisances/cross_conditional_outcome.py:
```diff
@@ -0,0 +1,224 @@
+
+from med_bench.utils.utils import _get_train_test_lists, _get_interactions
+
+def estimate_cross_conditional_mean_outcome_discrete(m, x, y, f, regressors):
```
not sure about a separate file from `estimate_conditional_mean_outcome`

bthirion replied:
Indeed, it would be fine having all  nuisances in one module IMHO

judithabk6 commented on src/med_bench/nuisances/density.py:
```diff
@@ -0,0 +1,315 @@
+
+    return f_t0, f_t1
+
+def estimate_mediator_probability(t, m, x, y, crossfit, clf_m,
```
we should harmonize and explain more propensity vs probability vs density

judithabk6 commented on src/med_bench/nuisances/utils.py:
```diff
@@ -0,0 +1,83 @@
+        reg = RidgeCV(alphas=alphas, cv=CV_FOLDS)
+    else:
+        reg = RandomForestRegressor(n_estimators=100, min_samples_leaf=10)
+
```
the objective of the refactor is to generalize to any method beyond ridge and forests, and not hardcode parameters

judithabk6 commented on src/med_bench/utils/scores.py:
```diff
@@ -0,0 +1,45 @@
+import numpy as np
```
needs docstrings and refs

judithabk6 reviewed (commented):
additional comment here, there is a whole `nuisances` folder and a nuisances.py files in utils, should be reorganized.

bthirion reviewed (commented):
LMK whenever you want a review

bthirion commented on src/med_bench/estimation/mediation_g_computation.py:
```diff
@@ -0,0 +1,64 @@
+
+        t, m, x, y = self._resize(t, m, x, y)
+
+        mu_0mx, mu_1mx, y0m0, y0m1, y1m0, y1m1 = self._estimate_cross_conditional_mean_outcome_nesting(
```
avoid very long lines

houssamzenati replied:
This is addressed in commit 6378d4318af2c23ec76999d758ad47475dd26d6e

bthirion commented on src/med_bench/estimation/mediation_tmle.py:
```diff
@@ -0,0 +1,207 @@
+
+        # reg_cross.fit(x, (mu_t1_mx_star).squeeze())
+
+        # omega_t = reg_cross.predict(x)
```
Please remove unused lines

bthirion commented on src/med_bench/example.py:
```diff
@@ -0,0 +1,119 @@
+from med_bench.utils.constants import CV_FOLDS
+
+
+if __name__ == "__main__":
```
I don't see the point of having this in an example.

bthirion commented on src/med_bench/example.py:
```diff
@@ -0,0 +1,119 @@
+    results_df = pd.DataFrame(results)
+
+    # Display or save the DataFrame
+    print(results_df)
```
I'd be in favor of plotting the results.

bthirion commented on src/med_bench/nuisances/density.py:
```diff
@@ -0,0 +1,315 @@
+    f_m1x, array-like, shape (n_samples)
+        probabilities f(M|T=1,X)
+    """
+    # if not is_array_integer(m):
```
remove unused lines

houssamzenati replied:
This is addressed in commit 6378d4318af2c23ec76999d758ad47475dd26d6e

bthirion reviewed (commented):
Thx ! 
This looks very nice. Nit-picking on details.

brash6 commented:
I feel from @bthirion comments that : 

- A review from @houssamzenati and @judithabk6 can also be useful 
- A meeting could also enable us to precise what works as expected and what has to be changed

judithabk6 commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,526 @@
+
+    def _fit_cross_conditional_mean_outcome_nuisance_discrete(self, t, m, x, y):
+        """
+        Fits the cross conditional mean outcome E[E[Y|T=t,M,X]|T=t',X] discrete
```
should be binary, for now not implemented for any discrete mediator

judithabk6 commented:
Actually, there is a bit of a blind spot in the PR as it currently stands, which is that the old and new implementations coexist. This is what Houssam started to correct in the example.py file, but from what I understood, the goal was to have an example code to compare the old and new. For me, when the PR is merged, this file should not exist and should be integrated into the docs (or exist temporarily and be deleted once we have integrated this new API into the documentation PR #87).
But currently all the tests run on the old implementation of the estimators.
What I am going to do now (I start, I hope I have time to finish this by the end of the week) is to replace the estimators in the function `get_estimation` by the new ones to run my experiments and compare the results to check that everything is ok. However I think that what we want to do ultimately is to delete entirely the files `get_estimation.py` and `mediation.py`. This means that we have to rethink the tests (and maybe instantiate each combination that we want to test in the test code). What do you think? That's a lot of work, but in the spirit of allowing any scikit learn estimator, it does not make sense to hard code a few limited combinations in `get_estimation.py` IMHO

bthirion commented:
Indeed, we need to get rid of the duplicates. It's OK to complete remove get_estimation.py

bthirion commented:
I can take a bit of time to discuss that on Tuesday afternoon (>4.30pm)

houssamzenati commented:
I am currently working towards adapting the tests to the new modularized estimators. 

In the `get_estimation.py` file I however found that sometimes there might problems in terms of naming for the options tested "elif estimator == "mediation_ipw_reg_calibration":, : see lines https://github.com/judithabk6/med_bench/blob/ae48827593177eb9977199cd9dcf6e301b15b619/src/med_bench/get_estimation.py#L128 no calibration is used (`calibration=None`) but right after that "elif estimator == "mediation_ipw_reg_calibration_cf":" is tested and this time calibration is set to "sigmoid" in line https://github.com/judithabk6/med_bench/blob/ae48827593177eb9977199cd9dcf6e301b15b619/src/med_bench/get_estimation.py#L154

same issue lines https://github.com/judithabk6/med_bench/blob/ae48827593177eb9977199cd9dcf6e301b15b619/src/med_bench/get_estimation.py#L206 for the G-computation as well see line https://github.com/judithabk6/med_bench/blob/ae48827593177eb9977199cd9dcf6e301b15b619/src/med_bench/get_estimation.py#L310

starting from this line https://github.com/judithabk6/med_bench/blob/ae48827593177eb9977199cd9dcf6e301b15b619/src/med_bench/get_estimation.py#L388

the naming of the options tested w.r.t calibration however is logical with what is tested.

houssamzenati commented:
Following our discussions @judithabk6 here is a summary of the cross-fitting issue and what is related to the tests (which you have deemed to be the priority, and have also agreed to, with you):

- many of the tests, and even Hadrien's file `example.py` (which is a test) actually use the cross-fitting and I believed at first it was necessary to make sure that the tests pass to revert to the old cross-fitting methods before merging this PR, from what I understood it is better to "deactivate" those tests and merge this PR once we ensure all the tests without cross-fitting run perfectly : therefore it is not the priority now to revert to the cross-fitting

- my idea for the cross-fitting syntax was the following, instead of having:
```python
estimator = ImportanceWeighting(
            clip=1e-6, trim=0, regressor=reg, classifier=clf
        )
estimator.fit(t, m, x, y)
causal_effects = estimator.estimate(t, m, x, y)
```

we could have directly

```python
estimator = ImportanceWeighting(
            clip=1e-6, trim=0, regressor=reg, classifier=clf
        )
causal_effects = estimator.estimate_crossfit(t, m, x, y)
```

and inside the `med_bench/estimation/base.py` file, we could implement methods `_estimate_cross_fit_mediator_probability`, `_estimate_cross_fit_treatment_propensity_x`, `_estimate_cross_fit_treatment_propensity_xm`, `_estimate_cross_fit_conditional_mean_outcome(self, t, m, x, y)`, `_estimate_cross_fit_cross_conditional_mean_outcome(self, t, m, x, y)` that would be **EXACTLY** the same functions as before, in the version of the code we want to merge to.

@judithabk6, you asked why it was not possible to use the easy hack 

```python
cf_iterator = KFold(k=5)
for data_train, data_test in cf_iterator:
    result.append(DML(...., cross_fitting=False)
         .fit(train_data.X, train_data.t, train_data.m, train_data.y)\
         .estimate(test_data.X, test_data.t, test_data.m, test_data.y))
np.mean(result)
```
I just wanted to say that this sample splitting procedure is not the same as the one you originally had in terms of syntax, which was rather (for example for the treatment propensities `p_x` and `p_xm`):

```python
    train_test_list = _get_train_test_lists(crossfit, n, x)

    xm = np.hstack((x, m))

    for train_index, test_index in train_test_list:

        # p_x, p_xm model fitting
        clf_t_x = clf_t_x.fit(x[train_index, :], t[train_index])
        clf_t_xm = clf_t_xm.fit(xm[train_index, :], t[train_index])

        # predict P(T=1|X), P(T=1|X, M)
        p_x[test_index] = clf_t_x.predict_proba(x[test_index, :])[:, 1]
        p_xm[test_index] = clf_t_xm.predict_proba(xm[test_index, :])[:, 1]
```

Therefore since the splitting is not the same, the tests which compare to the stored values in the csv file will fail. We may adapt the easy hack above to do the same splits, or only revert to this old code, with the new two syntax I have proposed.

1) When doing model selection in causal inference, as Doutreligne did: `estimator.fit()` on the train samples and to perform model selection, `estimator.estimate()` on the evaluation samples 
2) OR without model selection and the old way, with cross-fitting, directly `estimator.estimate_cross_fit()` without fitting on training samples.

brash6 commented:
I am available for a call on Tuesday afternoon (>4.30pm) 
Just like this comment if you are too @bthirion @judithabk6 @houssamzenati

bthirion commented:
> Following our discussions @judithabk6 here is a summary of the cross-fitting issue and what is related to the tests (which you have deemed to be the priority, and have also agreed to, with you):
> 
>     * many of the tests, and even Hadrien's file `example.py` (which is a test) actually use the cross-fitting and I believed at first it was necessary to make sure that the tests pass to revert to the old cross-fitting methods before merging this PR, from what I understood it is better to "deactivate" those tests and merge this PR once we ensure all the tests without cross-fitting run perfectly : therefore it is not the priority now to revert to the cross-fitting
> 
>     * my idea for the cross-fitting syntax was the following, instead of having:
> 
> 
> ```python
> estimator = ImportanceWeighting(
>             clip=1e-6, trim=0, regressor=reg, classifier=clf
>         )
> estimator.fit(t, m, x, y)
> causal_effects = estimator.estimate(t, m, x, y)
> ```
> 
> we could have directly
> 
> ```python
> estimator = ImportanceWeighting(
>             clip=1e-6, trim=0, regressor=reg, classifier=clf
>         )
> causal_effects = estimator.estimate_crossfit(t, m, x, y)
> ```
> 
> and inside the `med_bench/estimation/base.py` file, we could implement methods `_estimate_cross_fit_mediator_probability`, `_estimate_cross_fit_treatment_propensity_x`, `_estimate_cross_fit_treatment_propensity_xm`, `_estimate_cross_fit_conditional_mean_outcome(self, t, m, x, y)`, `_estimate_cross_fit_cross_conditional_mean_outcome(self, t, m, x, y)` that would be **EXACTLY** the same functions as before, in the version of the code we want to merge to.
> 
> @judithabk6, you asked why it was not possible to use the easy hack
> 
> ```python
> cf_iterator = KFold(k=5)
> for data_train, data_test in cf_iterator:
>     result.append(DML(...., cross_fitting=False)
>          .fit(train_data.X, train_data.t, train_data.m, train_data.y)\
>          .estimate(test_data.X, test_data.t, test_data.m, test_data.y))
> np.mean(result)
> ```
> 
> I just wanted to say that this sample splitting procedure is not the same as the one you originally had in terms of syntax, which was rather (for example for the treatment propensities `p_x` and `p_xm`):
> 
> ```python
>     train_test_list = _get_train_test_lists(crossfit, n, x)
> 
>     xm = np.hstack((x, m))
> 
>     for train_index, test_index in train_test_list:
> 
>         # p_x, p_xm model fitting
>         clf_t_x = clf_t_x.fit(x[train_index, :], t[train_index])
>         clf_t_xm = clf_t_xm.fit(xm[train_index, :], t[train_index])
> 
>         # predict P(T=1|X), P(T=1|X, M)
>         p_x[test_index] = clf_t_x.predict_proba(x[test_index, :])[:, 1]
>         p_xm[test_index] = clf_t_xm.predict_proba(xm[test_index, :])[:, 1]
> ```
> 
> Therefore since the splitting is not the same, the tests which compare to the stored values in the csv file will fail. We may adapt the easy hack above to do the same splits, or only revert to this old code, with the new two syntax I have proposed.
> 
>     1. When doing model selection in causal inference, as Doutreligne did: `estimator.fit()` on the train samples and to perform model selection, `estimator.estimate()` on the evaluation samples
> 
>     2. OR without model selection and the old way, with cross-fitting, directly `estimator.estimate_cross_fit()` without fitting on training samples.

I 'm not obsessed by reproducing exactly previous values, if the cross-fitting is changed. What matters is that the results is sound.

judithabk6 commented:
Hi,
A little summary of today's meeting
- [ ] option `cross_fitting=True` in the estimator should raise a nonImplementedError, and it should be testes when requesting a cross-fitted estimator (in the updated test suite)
- [ ] change GComputation to have the explicit summation on m in {0, 1} when the mediator is binary, and it continuous the current version
- [ ] check whether the mediator is binary (estimators `simulation_based`, `mediation_g_estimator` only accept a binary mediator, the rest accept all types of mediators, and `r_mediation_dml` doesn't accept only one covariate)
- [ ] merge and refactor the tests
- [ ] also `ImportanceWeighting` should be renamed InversePropensityWeighting

let me know if you think of anything else or need anything @brash6! thanks!!!! <3

judithabk6 commented:
or we implement a "pseudo cross-fitting" like this (for instance for IPW). It does not give the same results, I don't know...., it's not that bad, but this would be temporary until we re-implement the cross-fitting
```python
    elif estimator == "mediation_ipw_noreg_cf":
        effects = mediation_IPW(
            y,
            t,
            m,
            x,
            trim=0,
            regularization=False,
            forest=False,
            crossfit=2,
            clip=1e-6,
            calibration=None,
        )
        cs, alphas = _get_regularization_parameters(regularization=False)
        clf = LogisticRegressionCV(random_state=42, Cs=cs, cv=CV_FOLDS)
        reg = RidgeCV(alphas=alphas, cv=CV_FOLDS)
        estimator = ImportanceWeighting(
            clip=1e-6, trim=0, regressor=reg, classifier=clf
        )
        result = list()
        train_test_list = _get_train_test_lists(crossfit=2, n=len(t), x=x)
        for train_index, test_index in train_test_list:
            estimator.fit(t[train_index],
                          m[train_index, :],
                          x[train_index, :],
                          y[train_index])
            causal_effects = estimator.estimate(t[test_index],
                                                m[test_index, :],
                                                x[test_index, :],
                                                y[test_index])
            result.append(transform_outputs(causal_effects))
        effects = np.array(result).mean(axis=0)
        
        ````

bthirion commented:
I think that our long-term intention is to introduce proper cross-fitting. Maybe we should not introduce an intermediate version if does not do exactly what we want ultimately. Also, I would in all case defer that to a future PR.

judithabk6 commented:
sounds fair! agreed! :)

judithabk6 commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,496 @@
+        """Fits the nuisance parameter for the density f(M=m|T, X)"""
+        # estimate mediator densities
+        clf_param_grid = {}
+        classifier_m = GridSearchCV(self.classifier, clf_param_grid)
```
I don't think we need this step. something like
```
classifier_m = clone(self.classifier)
```
?
true at many places, todo now or in the next PR?

judithabk6 commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,496 @@
+        t0_x = np.hstack([t0.reshape(-1, 1), x])
+        t1_x = np.hstack([t1.reshape(-1, 1), x])
+
+        fm_0 = self._classifier_m.predict_proba(t0_x)[:, 1]
```
it seems false, for me it predicts P(M=1|T=0, X=x) and not P(M=m|T=0, X=x), should be
```
fm_0 = self._classifier_m.predict_proba(t0_x)[:, m]
```

no? and same at the line after

judithabk6 commented on src/med_bench/estimation/mediation_g_computation.py:
```diff
@@ -0,0 +1,104 @@
+        """Fits nuisance parameters to data"""
+        t, m, x, y = self._resize(t, m, x, y)
+
+        if is_array_integer(m):
```
this if should test whether m is binary, not an integer array.
```
if len(np.unique(m))==2:
```

judithabk6 commented on src/med_bench/estimation/mediation_dml.py:
```diff
@@ -0,0 +1,117 @@
+            Regressor used for mu estimation, can be any object with a fit and predict method
+        classifier 
+            Classifier used for propensity estimation, can be any object with a fit and predict_proba method
+        clips : float
```
not a parameter here, to remove, same for trim

judithabk6 reviewed (commented):
I focused on the main things changed since last time.
A quick thing to do before merging would be to think about the naming of methods in base.py, and at least update the docstrings

also, it seems that the tests have not run since those commits, anyone knows why?

brash6 commented:
> Hi, A little summary of today's meeting
> 
> * [ ]  option `cross_fitting=True` in the estimator should raise a nonImplementedError, and it should be testes when requesting a cross-fitted estimator (in the updated test suite)
> * [x]  change GComputation to have the explicit summation on m in {0, 1} when the mediator is binary, and it continuous the current version (this would fix [Adding the continuous use case to the G Computation implementation #97](https://github.com/judithabk6/med_bench/issues/97) I think)
> * [ ]  check whether the mediator is binary (estimators `simulation_based`, `mediation_g_estimator` only accept a binary mediator, the rest accept all types of mediators, and `r_mediation_dml` doesn't accept only one covariate)
> * [x]  merge and refactor the tests
> * [x]  also `ImportanceWeighting` should be renamed InversePropensityWeighting
> * [x]  have input regressor and classifier only when needed
> 
> let me know if you think of anything else or need anything @brash6! thanks!!!! <3

I have implemented almost all the points listed here except : 
- I haven't changed the estimators running the cross fitting on the old version in the test suites, happy to confirm what we want to do tomorrow with this
- The tests may still need to have the check on the nature of the input mediator 
- I haven't really understood what to do with the estimators that don't have an equivalent in the new refactored version
- What do we do with the TMLE ?

judithabk6 commented:
> * I haven't changed the estimators running the cross fitting on the old version in the test suites, happy to confirm what we want to do tomorrow with this
yes! we can drop that (cf Bertrand's comment above)
> * The tests may still need to have the check on the nature of the input mediator
ok. Regarding this, we discussed getting rid of the `get_estimation` and replace it by a sort of big dictionary in the test file, is it doable? otherwise maybe we should move the `get_estimation` function to the test file to indicate it should only be used there?
> * I haven't really understood what to do with the estimators that don't have an equivalent in the new refactored version
we should keep them in mediation.py and use them as they are
> * What do we do with the TMLE ?
not sure, it is not tested, and I haven't reviewed it a lot. Maybe add TMLE in the tolerance test, with the smallest possible tolerance that works?
> * Let's talk about your other review comments tomorrow

brash6 commented:
@judithabk6 @bthirion 
We just addressed all the comments from Judith with @houssamzenati, however, some issues still remain : 
- `GComputation` estimator still has large discrepencies with expected values in the tolerance tests : Do we increase the tolerance thresholds or we try to understand the problem deeper ? 
- TMLE is still not tested
- Documentations has to be enhanced (docstrings and online doc)
- Some optimizations can still be performed in a further PR (DML removal, some functions merging, etc.)

bthirion commented:
FOr GComputation, ideally, we should localize the issue. If it takes too long, merge as is but open in parallel an issue together with a code that outllines this issue so that we can investigate it later.

bthirion commented:
can't we have a test on TMLE similar to that of MultiplyRobust (actually the same test ?)

bthirion commented:
For code and doc improvements, the good way to proceed is to open a bunch of issues so that we don't lose track of the post-refactoring changes.

judithabk6 commented:
@bthirion I'm on it, I agree with you that more exploration is necessary to figure out exactly what is going on with Gcomputation, and add tests for TMLE, at least, and improve the docstring and naming of the methods of the base estimator

houssamzenati reviewed (commented):
The current version seems ready to my mind to be merged. Should we do so and move on to the documentation?

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,351 @@
+        train0 = train[t[train] == 0]
+
+        train_mean, train_nested = np.array_split(train, 2)
+        # train_mean = train
```
```suggestion
```

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,351 @@
+
+        train_mean, train_nested = np.array_split(train, 2)
+        # train_mean = train
+        # train_nested = train
```
```suggestion

```

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,351 @@
+        p_x : array-like, shape (n_samples)
+            probabilities P(T=1|X)
+        """
+        # predict P(T=1|X), P(T=1|X, M)
```
```suggestion

```

bthirion commented on src/med_bench/estimation/base.py:
```diff
@@ -0,0 +1,351 @@
+        """
+        xm = np.hstack((x, m))
+
+        # predict P(T=1|X), P(T=1|X, M)
```
```suggestion

```

bthirion commented on src/tests/estimation/test_get_estimation.py:
```diff
@@ -110,24 +108,28 @@ def effects_chap(x, t, m, y, estimator, config):
 
 def test_tolerance(effects, effects_chap, tolerance):
     error = abs((effects_chap - effects) / effects)
+    # print(error)
```
```suggestion

```

bthirion reviewed (approved):
LGTM

bthirion commented:
> The current version seems ready to my mind to be merged. Should we do so and move on to the documentation?

+1 for merging now.
